src/appleseed/foundation/meta/tests/test_imagefilewriter.cpp

```diff
@@ -0,0 +1,245 @@
+            
+            image.clear(Color4b(50, 100, 150, 42));
+            
+            GenericImageFileWriter writer(ImageFilePath);
+            writer.append_image(&image);
+            writer.write();
+        }
+        
+        {
+            GenericImageFileReader reader;
+            std::unique_ptr<Image> image(reader.read(ImageFilePath));
+
+            for (size_t y = 0; y < 2; ++y)
+            {
+                for (size_t x = 0; x < 2; ++x)
+                {
+                    Color4b c;
+                    image->get_pixel(x, y, c);
+                    EXPECT_EQ(Color4b(50, 100, 150, 42), c);
+                }
+            }
+        }
+    }
+    
+    TEST_CASE(WriteHDR_CorrectlyWritesImagePixels)
+    {
+        const char* ImageFilePath = "unit tests/outputs/test_imagefilewriter.hdr";
+        
+        {
+            Image image(2, 2, 2, 2, 3, PixelFormatUInt16);
+            
+            image.clear(Color3b(50, 100, 150));
+            
+            GenericImageFileWriter writer(ImageFilePath);
+            writer.append_image(&image);
+            writer.write();
+        }
+        
+        {
+            GenericImageFileReader reader;
+            std::unique_ptr<Image> image(reader.read(ImageFilePath));
+
+            for (size_t y = 0; y < 2; ++y)
+            {
+                for (size_t x = 0; x < 2; ++x)
+                {
+                    Color3b c;
+                    image->get_pixel(x, y, c);
+                    EXPECT_EQ(Color3b(50, 100, 150), c);
+                }
+            }
+        }
+    }
+    
+    TEST_CASE(WriteJPG_CorrectlyWritesImagePixels)
+    {
+        const char* ImageFilePath = "unit tests/outputs/test_imagefilewriter.jpg";
+        
+        {
+            Image image(2, 2, 2, 2, 3, PixelFormatFloat);
+            
+            image.clear(Color3f(0.6f, 0.7f, 0.8f));
+            
+            GenericImageFileWriter writer(ImageFilePath);
+            writer.append_image(&image);
+            writer.write();
+        }
+        
+        {
+            GenericImageFileReader reader;
+            std::unique_ptr<Image> image(reader.read(ImageFilePath));
+
+            for (size_t y = 0; y < 2; ++y)
+            {
+                for (size_t x = 0; x < 2; ++x)
+                {
+                    Color3f c;
+                    image->get_pixel(x, y, c);
+                    EXPECT_FEQ_EPS(Color3f(0.6f, 0.7f, 0.8f), c, 0.003f);
+                }
+            }
+        }
+    }
+    
+    TEST_CASE(WritePNG_CorrectlyWritesImagePixels)
+    {
+        const char* ImageFilePath = "unit tests/outputs/test_imagefilewriter.png";
+        
+        {
+            Image image(2, 2, 2, 2, 4, PixelFormatFloat);
+            Color4b color(50, 100, 150, 42);
+            color.unpremultiply_in_place();
+            color.rgb() = fast_linear_rgb_to_srgb(color.rgb());
+            color = saturate(color);
+            color.premultiply_in_place();
```

Comment:
Can you explain why you do this only for `png` and `tiff` ?

Reply:
![error_PNG](https://user-images.githubusercontent.com/55926883/75180737-cf740c80-570a-11ea-8e03-ace09ef7ec48.png)
Without these few lines, the actual written value will have small discrepancy from expected value. So far I could not figure out a plausible cause, so I thought it could be a colorspace issue. (however, colorspace conversion should already be handled by other source code).

Reply:
I think you should add a comment in the code to explain this.

Reply:
That doesn't make sense. PNG is lossless, you should be able to read back exactly what you wrote. The problem is that you're now using a floating point image (`PixelFormatFloat`) with a format that doesn't support floating point values.

Reply:
PixelFormatUInt8 & PixelFormatUInt16 image do not work for PNG. The testcase "WritePNG_CorrectlyWritesPixels_pixelB_colorB" used PixelFormatUInt8. 
```
TEST_CASE(WritePNG_CorrectlyWritesImagePixels_pixelB_colorB)
    {
        const char* ImageFilePath = "unit tests/outputs/test_imagefilewriter_pixelB_colorB.png";
        
        {
            Image image(2, 2, 2, 2, 4, PixelFormatUInt8);
            
            image.clear(Color4b(50, 100, 150, 42));
            
            GenericImageFileWriter writer(ImageFilePath);
            writer.append_image(&image);
            writer.write();
        }
        
        {
            GenericImageFileReader reader;
            std::unique_ptr<Image> image(reader.read(ImageFilePath));

            for (size_t y = 0; y < 2; ++y)
            {
                for (size_t x = 0; x < 2; ++x)
                {
                    Color4b c;
                    image->get_pixel(x, y, c);
                    EXPECT_EQ(Color4b(50, 100, 150, 42), c);
                }
            }
        }
    }
```

Reply:
Here's the source files that I wrote for the experiment: https://github.com/ChinYing-Li/Foundation_Image_IO/tree/master/debug

I will change PixelFormatUInt16 to UInt8 as necessary and upload the results soon.